vivarium/core/registry.py

```diff
@@ -335,130 +363,73 @@ def divide_null(state):
+        :py:class:`bson.codec_options.TypeCodec` that are used by PyMongo to
+        serialize and (optionally) deserialize data.
         """
-        raise NotImplementedError(
-            f'{self} has not implemented serialize_to_string().')
-
+        return []
+    
     def serialize(self, data):
-        """Serialize some data.
-
-        Subclasses that are serializing to strings can just implement
-        :py:meth:`vivarium.core.registry.Serializer.serialize_to_string`,
-        and this method will use that implementation to generate a
-        string that serializes the data.
-
-        Subclasses with more complex serialization needs must override
-        this function.
-
-        Args:
-            data: Data to serialize.
-
-        Returns:
-            The serialized data.
+        """Serialize data using correct codec.
+        
+        This is typically only called in the case that individual stores
+        are assigned custom serializers. For maximum performance, serialization
+        should be left to PyMongo instead of calling this function.
         """
-        string_serialization = self.serialize_to_string(data)
-        if not isinstance(string_serialization, str):
-            raise ValueError(
-                f'{self}.serialize_to_string() returned invalid '
-                f'serialization: {string_serialization}')
-
-        return f'!{self.name}[{string_serialization}]'
-
-    def can_serialize(self, data):
-        """Check whether the serializer can serialize some data.
-
-        Every subclass must implement this method.
-        """
-        raise NotImplementedError(
-            f'{self} has not implemented can_serialize().')
-
-    def deserialize_from_string(self, data):
-        """Deserialize data from a string.
-
-        This function is the inverse of
-        :py:meth:`vivarium.core.registry.Serializer.serialize_to_string`.
-        It need only be overridden in a subclass if that subclass does
-        not override
-        :py:meth:`vivarium.core.registry.Serializer.deserialize`.
-        """
-        raise NotImplementedError(
-            f'{self} has not implemented deserialize_from_string().')
+        for codec in self.codecs:
+            if isinstance(data, codec.python_type):
+                return codec.transform_python(data)
 
     def deserialize(self, data):
-        """Deserialize some data.
-
-        Subclasses relying on
-        :py:meth:`vivarium.core.registry.Serializer.serialize_to_string`
-        should not override this function and instead override
-        :py:meth:`vivarium.core.registry.Serializer.deserialize_from_string`.
-
-        Args:
-            data: Serialized data to deserialize.
-
-        Returns:
-            The deserialized data.
+        """Given a string of the form ``!...[data here]``,
+        where ``...`` is ``self.name``, this returns the 
+        string inside the square brackets.
```

Comment:
Is the idea that serializers that don't use PyMongo codecs should override this function with their own deserialization logic? If so, that should be explained in the docstring

Reply:
That's right. This method should only be overridden if the user also overrode `serialize` to enable custom serialization on a store-by-store basis (as opposed to all data of a specific type being serialized the same way). 

As an aside, the `TypeCodec` system allows users to define the `bson_type` attribute and `transform_bson` method to deserialize all BSON types the same way. Letting PyMongo perform all serialization and deserialization would be fastest. Unfortunately, our serializers do not represent a one-to-one mapping between python and BSON types (e.g. lists and arrays both serialize to lists), which is why a lot of them override the `deserialize` method and use the slow `deserialize_value` API.